### PR TITLE
feat: yarn2 lockfile support

### PR DESCRIPTION
**Pull request by christophehurpeau**
Support yarn 2 lockfiles

See https://github.com/renovatebot/renovate/issues/5230#issuecomment-612887676

### CODE REVIEW

**Comment by christophehurpeau** [comment]
There is something I didn't understand : what is `packageFile.yarnIntegrity`. With this pr, for yarn2 it will always be false. Is-it a problem ?

File: lib/manager/npm/extract/yarn.ts [hunk 1]
```diff
@@ -7,30 +7,77 @@ export type YarnLock = Record<string, string> & {
...
 };
 
+interface ParseRangeReturnType {
+  protocol: string | null;
+  source: string | null;
+  selector: string | null;
+  params: string | null;
+}
+
+// https:// github.com/yarnpkg/berry/blob/b128988379a1f257b2406ebf5058b67861403bd4/packages/yarnpkg-core/sources/structUtils.ts#L260-L303
```
Comment (JamieMagee) [inline lib/manager/npm/extract/yarn.ts:17]: I understand why you would copy a couple of functions, instead of depending on `@yarnpkg/core`, but I'm not sure it complies with the terms of the license. `@yarnpkg/core` is [licensed under the BSD 2-clause license](https://github.com/yarnpkg/berry/blob/master/LICENSE.md), which includes the following bullet-point

>Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

**Comment by JamieMagee** [comment]
> There is something I didn't understand : what is `packageFile.yarnIntegrity`. With this pr, for yarn2 it will always be false. Is-it a problem ?

Yes, I think that would be an issue. At least if you want this released before Renovate v20.

For some context, Yarn added integrity hashes in version `1.10.0`. So if you generated a `yarn.lock` file using Yarn 1.10.0+ it would add the integrity property, but if you ran it with Yarn < 1.10.0 it would remove the integrity property.

The `yarnIntegrity` property was added to detect what version of Yarn Renovate should use. We also bundle a Yarn 1.9.4 binary, but we're planning to remove that for Yarn v20.

File: lib/manager/npm/extract/yarn.ts (hunk 1 above)
Comment (christophehurpeau) [reply lib/manager/npm/extract/yarn.ts:17]: oh I did not know we could use @yarnpkg/core directly ! I will make the changes :)

**Comment by christophehurpeau** [comment]
About yarnIntegrity so I guess it's not an issue ? The yarn used will be yarn 2 in the repository, no matter which yarn starts it

**Comment by JamieMagee** [comment]
I think it might be an issue. Here's the code we're talking about

```ts
// See if renovate is installed locally
      const installedPath = join(
        await getInstalledPath('yarn', {
          local: true,
        }),
        'bin/yarn.js'
      );
      cmd = `node ${installedPath}`;
      const yarnIntegrity =
        config.upgrades &&
        config.upgrades.some((upgrade) => upgrade.yarnIntegrity);
      if (!yarnIntegrity) {
        logger.warn('Using yarn@1.9.4 for install is deprecated');
        ...
        cmd = 'node ' + join(renovatePath, 'bin/yarn-1.9.4.js');
```

It doesn't matter what version yarn reports. If `yarnIntegrity` is always false then yarn 1.9.4 will be used, and we haven't run `yarn policies set-version berry` on it to use the version included in the repo. I'm not even sure if that would work, as it's such an old version.

**Comment by christophehurpeau** [comment]
@JamieMagee when you run `yarn policies set-version berry` it downloads yarn in the repository and create a config file with the yarn path. You only need to run it once, so renovate doesn't need to do it
Then yarn no matter the version (even 1.9) check this config file. If a yarn path is defined, it calls this yarn instead: https://github.com/yarnpkg/yarn/blob/1.9-stable/src/cli/index.js#L582-L600

**Comment by christophehurpeau** [comment]
Well in fact it seems that yarn 1.9 doesnt support yml, so maybe it doesnt work indeed (https://github.com/yarnpkg/yarn/blob/1.22-stable/src/rc.js#L58)

**Comment by JamieMagee** [comment]
From what I can see, with your latest commit we should be setting `packageFile.yarnIntegrity` correctly, right? Then we're just waiting for a fix for `@yarnpkg/core` from upstream?

**Comment by christophehurpeau** [comment]
@JamieMagee yes :)

**Comment by christophehurpeau** [comment]
@JamieMagee the yarn issue doesn't seem to be solving soon, so I suggest to disable .d.ts checks by typescript in tsconfig with `skipLibCheck`, what do you think ?

**Comment by viceice** [comment]
@christophehurpeau We can use `patch-package` to fix it locally until a new version is released.

**Comment by JamieMagee** [comment]
@christophehurpeau I don't think we're willing to remove `skipLibCheck`.

`patch-package` is definitely an option, as @viceice suggested, but it's brittle, and I'd like to wait a little bit first to see if we can get the fix in upstream. Yarn is iterating quickly on RCs right now, so I think there's a good chance it could be fixed. If they release `2.0.0`, and it's still broken, then we can fix it here.

**Comment by JamieMagee** [comment]
@christophehurpeau Looks like the issue we were waiting for is resolved in `@yarnpkg/core:2.0.0-rc.29`. If you have time can you deconflict this PR and bump the version? If not, let me know and I can do it.

**Comment by christophehurpeau** [comment]
Yes, I’ll have some time this week!
thank you

File: lib/manager/npm/post-update/yarn.ts [hunk 2]
```diff
@@ -46,15 +49,38 @@ export async function generateLockFile(
...
+    const configuration = await Configuration.find(
+      npath.toPortablePath(cwd),
+      null,
+      { strict: false }
+    ).catch(() => null);
+    const yarnPath = configuration && configuration.get('yarnPath');
+
+    const yarnVersion =
+      yarnPath && (await exec(`${yarnPath} --version`)).stdout;
+    const isYarn1 = !yarnVersion || semver.getMajor(yarnVersion) === 1;
```
Comment (christophehurpeau) [inline lib/manager/npm/post-update/yarn.ts:61]: There is an other way we could do this: checking the lockfile (if __metadata exists, it's yarn 2). This version prevent installing yarn 1 if we don't need it, but I'm not sure it's the right one ?

File: lib/manager/npm/post-update/yarn.ts [hunk 3]
```diff
@@ -46,15 +49,38 @@ export async function generateLockFile(
...
-    let installYarn = 'npm i -g yarn';
-    const yarnCompatibility = config.compatibility?.yarn;
-    if (validRange(yarnCompatibility)) {
-      installYarn += `@${quote(yarnCompatibility)}`;
+    const configuration = await Configuration.find(
+      npath.toPortablePath(cwd),
+      null,
+      { strict: false }
+    ).catch(() => null);
+    const yarnPath = configuration && configuration.get('yarnPath');
```
Comment (christophehurpeau) [inline lib/manager/npm/post-update/yarn.ts:57]: it seems the path is a fullpath, would that be a problem with exec ? it seems exec is done in docker ?

**Comment by christophehurpeau** [comment]
It seems this PR was removed: https://github.com/renovatebot/renovate/pull/5657 so I still need to add the tests back

File: lib/manager/npm/extract/yarn.ts [hunk 4]
```diff
@@ -8,23 +9,26 @@ export type YarnLock = Record<string, string>;
...
+    const parsed = parseSyml(yarnLockRaw);
     const lockFile: YarnLock = {};
 
-    for (const [entry, val] of Object.entries(
-      yarnLockParsed.object as LockFileEntry
-    )) {
-      logger.trace({ entry, version: val.version });
-      lockFile[entry] = val.version;
-    }
+    Object.keys(parsed).forEach((key) => {
```
Comment (viceice) [inline lib/manager/npm/extract/yarn.ts:15]: Can you please use a for of loop as before

File: lib/manager/npm/post-update/yarn.ts (hunk 3 above)
Comment (viceice) [reply lib/manager/npm/post-update/yarn.ts:57]: We should move the version extraction to the extract phase and use config.compatibility.yarn here

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/manager/npm/post-update/yarn.ts (hunk 3 above)
Comment (christophehurpeau) [reply lib/manager/npm/post-update/yarn.ts:57]: to clarify: yarnPath contains yarn 2 executable path witch is in the repo

Comment (viceice) [reply lib/manager/npm/post-update/yarn.ts:57]: I know, I mean to fetch the yarn version, which should be done at extraction phase.

**Comment by JamieMagee** [comment]
@christophehurpeau I rebased your changes on top of current master, and resolved the conflicts.

**Comment by viceice** [comment]
needs coverage fixes

https://codecov.io/gh/renovatebot/renovate/pull/6045/src/lib/manager/npm/post-update/yarn.ts?before=lib/manager/npm/post-update/yarn.ts

https://codecov.io/gh/renovatebot/renovate/pull/6045/src/lib/util/fs/proxies.ts

**Comment by JamieMagee** [comment]
The difficult part here is the Mock of `Configuration.find`. @christophehurpeau I had a quick look, but couldn't mock it successfully. If you have time, can you take a look?

**Comment by christophehurpeau** [comment]
@JamieMagee Configuration.find is only useful if we don't want to install yarn, still waiting for a response on https://github.com/renovatebot/renovate/pull/6045#discussion_r442777914 ?

**Comment by christophehurpeau** [comment]
I'm probably not very clear, sorry :/
With yarn 2, yarn is most likely installed and gitted in the repository (in .yarn/releases/). Using Configuration.find allows to check that and directly use this yarn version.
However, we can still use yarn v1 installed globaly, in which case yarn v1 will look for the config file and run yarn v2 with the args we passed.
We still need to detect which version we use (args are different), but if we use yarn v2 we can check the version by checking if `__metadata` exists in the lockfile.

**Comment by viceice** [comment]
Isn't it enough to check for yarn v2 (version can be ignored, as yarn v2 is installed in repo) in extract phase.

Then in update phase we install global yarn (latest v1 if v2 is found) as now as we simply change some required args as needed.

File: lib/manager/npm/post-update/yarn.ts (hunk 2 above)
Comment (viceice) [reply lib/manager/npm/post-update/yarn.ts:62]: I would prefer to continue install global yarn (eg install latest if v2 is requested) and call thatwith the new args / env vars

**Comment by rarkins** [comment]
Yes, ideally we can identify the yarn version during extract and have a fairly simple artifacts updating

**Comment by christophehurpeau** [comment]
Okay I'll try this Thanks

File: lib/manager/npm/post-update/yarn.spec.ts [hunk 5]
```diff
@@ -29,7 +29,7 @@ describe(getName(__filename), () => {
     jest.resetAllMocks();
     env.getChildProcessEnv.mockReturnValue(envMock.basic);
   });
-  it.each([['1.22.0']])(
+  it.each([['1.22.0'], ['2.1.0']])(
```
Comment (christophehurpeau) [inline lib/manager/npm/post-update/yarn.spec.ts:32]: how to properly reset fs.readFile ? it's currently line 28 but doesn't cover this each

File: lib/manager/npm/post-update/yarn.ts [hunk 6]
```diff
@@ -1,13 +1,13 @@
 import is from '@sindresorhus/is';
-import { validRange } from 'semver';
+import { readFile, remove } from 'fs-extra';
```
Comment (viceice) [inline lib/manager/npm/post-update/yarn.ts:2]: Why this change? Please use `lib/util/fs` instead.

Comment (christophehurpeau) [reply lib/manager/npm/post-update/yarn.ts:2]: sorry, probably a wrong merge

**Comment by greenhat616** [comment]
🤔 what could we do to help accelerate the progress of support?

**Comment by christophehurpeau** [comment]
@greenhat616 I still have an issue with `jest.resetAllMocks();` and `it.each` as mentionned https://github.com/renovatebot/renovate/pull/6045#discussion_r460385325, do you have any suggestions ?

**Comment by JamieMagee** [comment]
@christophehurpeau Looking now

**Comment by JamieMagee** [comment]
Okay, I've got this working now. Some of the main changes were:
- `semver.major` doesn't work with ranges. So `const isYarn1 = !isValidYarnRange || major(yarnCompatibility) === 1;` was throwing an exception. I've changed this to use `minVersion(yarnCompatibility).major` instead
- Switched out `fs-extra` for `lib/util/fs` instead, and `readFile` for `readLocalFile`
- `readLocalFile` is only used once for yarn v2, but twice for yarn v2 (in `hasYarnOfflineMirror`) so I added an extra parameter to the tests to account for this

Will push in a few mins.

**Comment by JamieMagee** [comment]
I'm rebasing this branch locally, and will push out a deconflicted branch shortly.

**Comment by JamieMagee** [comment]
Just needs coverage for `remove` in `lib/util/fs/proxies`.

**Comment by christophehurpeau** [comment]
Thank you ! I added a test for `remove` as mentioned, should now be ready for review :)

File: lib/util/fs/proxies.spec.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+/* env jest */
```
Comment (viceice) [inline lib/util/fs/proxies.spec.ts:1]:
```suggestion
```

this is already configured globaly

**Review by viceice — DISMISSED** [review]
In a future pr we can improve by trying to read the version from `.yarnrc.yml`

https://github.com/itaylor/bucket-ts/blob/6719942c25791bcc1c16b0b6c9c53920ed1e6593/.yarnrc.yml#L5

File: package.json [hunk 8]
```diff
@@ -111,8 +111,9 @@
     "@renovate/pep440": "0.4.1",
     "@renovatebot/ruby-semver": "0.2.1",
     "@sindresorhus/is": "3.1.0",
-    "@yarnpkg/lockfile": "1.1.0",
     "aws-sdk": "2.725.0",
+    "@yarnpkg/core": "2.1.1",
+    "@yarnpkg/parsers": "2.1.0",
```
Comment (christophehurpeau) [inline package.json:116]:
```suggestion
    "@yarnpkg/core": "2.1.1",
    "@yarnpkg/parsers": "2.1.0",
    "aws-sdk": "2.725.0",
```

**Review by JamieMagee — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Review by greenhat616 — APPROVED** [review]

**Comment by christophehurpeau** [comment]
Thanks for merging ! I tested on a repository, and it seems that `--ignore-platform` is not supported, like `--ignore-script` (maybe it's the case for --ignore-engines too ?)

```
Unknown Syntax Error: Unsupported option name (\"--ignore-platform\").
```

**Comment by viceice** [comment]
looks like most options are removed on v2

https://yarnpkg.com/cli/install
